src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManagerImpl.java

```diff
@@ -0,0 +1,227 @@
+package net.neoforged.neoforge.transfer.transaction;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+
+@ApiStatus.Internal
+public class TransactionManagerImpl {
+    public static final ThreadLocal<TransactionManagerImpl> MANAGERS = ThreadLocal.withInitial(TransactionManagerImpl::new);
+
+    private final Thread thread = Thread.currentThread();
+    private final ArrayList<TransactionImpl> stack = new ArrayList<>();
+    private final ArrayList<Transaction.OuterCloseCallback> outerCloseCallbacks = new ArrayList<>();
+    private int currentDepth = -1;
+
+    public boolean isOpen() {
+        return currentDepth > -1;
+    }
+
+    public Transaction openOuter() {
+        if (isOpen()) {
+            throw new IllegalStateException("An outer transaction is already active on this thread.");
+        }
+
+        return open();
+    }
+
+    @Nullable
+    public TransactionContext getCurrentUnsafe() {
+        if (currentDepth == -1) {
+            return null;
+        } else if (stack.get(currentDepth).lifecycle == Transaction.Lifecycle.OPEN) {
+            return stack.get(currentDepth);
+        } else {
+            throw new IllegalStateException("May not call getCurrentUnsafe() from a close callback.");
+        }
+    }
+
+    /**
+     * Open a new transaction, outer or nested, without performing any state check.
+     */
+    Transaction open() {
+        currentDepth++;
+
+        if (stack.size() == currentDepth) {
+            stack.add(new TransactionImpl(currentDepth));
+        }
+
+        TransactionImpl current = stack.get(currentDepth);
+        current.lifecycle = Transaction.Lifecycle.OPEN;
+        return current;
+    }
+
+    void validateCurrentThread() {
+        if (Thread.currentThread() != thread) {
+            String errorMessage = String.format(
+                    "Attempted to access transaction state from thread %s, but this transaction is only valid on thread %s.",
+                    Thread.currentThread().getName(),
+                    thread.getName());
+            throw new IllegalStateException(errorMessage);
+        }
+    }
+
+    public Transaction.Lifecycle getLifecycle() {
+        if (currentDepth == -1) {
+            return Transaction.Lifecycle.NONE;
+        } else {
+            return stack.get(currentDepth).lifecycle;
+        }
+    }
+
+    private class TransactionImpl implements Transaction {
+        final int nestingDepth;
+        final ArrayList<CloseCallback> closeCallbacks = new ArrayList<>();
```

Comment:
Is there a reason this needs to be explicitly `ArrayList` instead of say just `List`?